pkg/cli/updater/resolver/resolver.go

```diff
@@ -181,33 +219,155 @@ func Resolve(ctx context.Context, token cfg.SecretData, c *Criteria) (*Version,
 		return nil, errors.Errorf("unknown channel %q", c.Channel)
 	}
 
-	v := getLatestVersion(versions[c.Channel])
-	if v == nil {
-		return nil, fmt.Errorf("no version found matching criteria")
+	return getLatestVersion(versions, c)
+}
+
+// stringInSlice returns true if the string is in the slice
+func stringInSlice(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+// getLatestSatisfyingConstraint returns the latest version that satisfies the constraint
+func getLatestSatisfyingConstraint(versions map[string][]Version, c *Criteria) (*Version, error) {
+	allowedChannels := []string{c.Channel}
+	if c.Channel != StableChannel {
+		// always allow stable channel versions
+		allowedChannels = append(allowedChannels, StableChannel)
 	}
 
-	// check if the stable version is greater than the latest version inside
-	// of our requested version, if so, use that.
-	//
-	// However, when using a mutable tag, do not automatically promote because
-	// we can't actually tell, reliably, when the version is greater, or lesser,
-	// than the stable version.
-	if c.Channel != StableChannel && !v.mutable {
-		stableV := getLatestVersion(versions[StableChannel])
-		if stableV != nil {
-			latestV := getLatestVersion([]Version{*v, *stableV})
-			if latestV != nil && latestV != v {
-				v = latestV
+	// we have a constraint, so find the latest version that satisfies it
+	constraints := make([]*semver.Constraints, len(c.Constraints))
+	for i, constraintStr := range c.Constraints {
+		// Limit the ability to use && and || in constraints so we can mutate the constraint
+		// to allow pre-releases
+		if strings.Contains(constraintStr, "&&") || strings.Contains(constraintStr, "||") {
+			return nil, errors.Errorf("multiple constraints within a single constraint are not supported")
+		}
+
+		// strip space from the constraint
+		constraintStr = strings.TrimSpace(constraintStr)
+		if constraintStr == "*" {
+			// for some reason * doesn't work with the prereleases hack we use below,
+			// so convert * into >= 0.0.0, it's equivalent.
+			constraintStr = ">=0.0.0"
+		}
+
+		// extract allowed channels from the constraint if there is one
+		exampleVer := oprStripRegex.ReplaceAllString(constraintStr, "$1")
```

Comment:
Why have a capture at all if you're just replacing a single character? Just replacing `^\D+` with `""` should suffice.

Reply:
Nuance of go regex, we want to ensure that there is a number after the non-digit, but in order to do that we have to match... which means it'll be replaced. So, we capture so we can add it back :/

Reply:
Oh, you want to ensure it's non-empty. Use a non-capturing group:
```
^\D(?:\d)
```

Reply:
Wait, nm - that'll still get matched. You want a zero-width lookahead, but it looks like go doesn't support that?

Reply:
Yep :(